### PR TITLE
With GPT2 tokenization, tokenize <|endoftext|> as a single token

### PR DESCRIPTION
**Pull request by gcooper-isi**
MegatronLM's GPT2-tokenization code breaks up "<|endoftext|>" into multiple tokens. However, in the English vocabulary file from GPT-2, there is a single token representing "<|endoftext|>". This pull request allows "<|endoftext|>" to be tokenized as a single token when it occurs as part of the text to be tokenized.

### CODE REVIEW

File: megatron/tokenizer/gpt2_tokenization.py [hunk 1]
```diff
@@ -233,15 +234,30 @@ def bpe(self, token):
         self.cache[token] = word
         return word
 
+    def separate_eot_tokens(self, text):
+        last_end_index = 0
+        for m in self.eot_pat.finditer(text):
```
Comment (wrichardson-isi) [inline megatron/tokenizer/gpt2_tokenization.py:239]: It looks like `self.eot_pat` is a string here, which causes the call to fail

Comment (gcooper-isi) [reply megatron/tokenizer/gpt2_tokenization.py:239]: Fixed

File: megatron/tokenizer/gpt2_tokenization.py [hunk 2]
```diff
@@ -233,15 +234,30 @@ def bpe(self, token):
...
             else:
-                token = ''.join(self.byte_encoder[b] for b in token.encode('utf-8'))
-            bpe_tokens.extend(bpe_token for bpe_token in self.bpe(token).split(' '))
+                tokens = re.findall(self.pat, segment)
+            for token in tokens:
+                if sys.version_info[0] == 2:
+                    token = ''.join(self.byte_encoder[ord(b)] for b in token)
+                else:
+                    token = ''.join(self.byte_encoder[b] for b in token.encode('utf-8'))
+                bpe_tokens.extend(bpe_token for bpe_token in self.bpe(token).split(' '))
```
Comment (wrichardson-isi) [inline megatron/tokenizer/gpt2_tokenization.py:260]: It looks to me like the `<|endoftext|>` tokens are separated from the rest of the text in the processing loop, but they are still being split into 7 different tokens in the final `bpe_tokens` list.

**Comment by gcooper-isi** [comment]
Closing, since `preprocess_data.py` can already add the end of text token (represented using a single number) after each training example. Also, if, for whatever reason, "<|endoftext|>" occurs inside the training example (e.g. from a web page talking about GPT-2), it's probably best for this not to be treated as actually ending the document.

**Pull request closed without merging**